### PR TITLE
[Feature] Custom Taxonomy population script [#OSF-7788]

### PR DESCRIPTION
**Pull request by mfraezz**
## Purpose
Add script to populate custom taxonomies when we have their information

## Changes
* Add custom manager for `Subject` model
* Add script

## Side effects
None expected

## Ticket
[[OSF-7788]](https://openscience.atlassian.net/browse/OSF-7788)

### CODE REVIEW

File: osf/models/subject.py [hunk 1]
```diff
@@ -1,25 +1,33 @@
...
+from include import IncludeQuerySet
 
 from website.util import api_v2_url
 
-from osf.models.base import BaseModel, ObjectIDMixin
+from osf.models.base import BaseModel, MODMCompatibilityQuerySet, ObjectIDMixin
 from osf.models.preprint_provider import PreprintProvider
 from osf.models.validators import validate_subject_hierarchy_length, validate_subject_provider_mapping
 
+class SubjectQuerySet(MODMCompatibilityQuerySet, IncludeQuerySet):
```
Comment (sloria) [inline osf/models/subject.py]: Does this need to inherit from IncludeQuerySet? I don't see any usages of .include anywhere.

File: osf/management/commands/populate_custom_taxonomies.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+            for s in hier:
+                preprint.subjects.add(s)
+        preprint.save()
+        preprint.reload()
+        new_hier = [s.object_hierarchy for s in preprint.subjects.exclude(children__in=preprint.subjects.all())]
+        logger.info('Successfully migrated preprint {}.\n\tOld hierarchy:{}\n\tNew hierarchy:{}'.format(preprint.id, old_hier, new_hier))
+
+def migrate(provider=None, data=None):
+    custom_provider = PreprintProvider.objects.filter(_id=provider).first()
+    assert custom_provider, 'Unable to find specified provider: {}'.format(provider)
```
Comment (mfraezz) [inline osf/management/commands/populate_custom_taxonomies.py]: I know, I chose this way to provide a more useful error message.

File: osf/models/subject.py (hunk 1 above)
Comment (mfraezz) [inline osf/models/subject.py]: Not at present, no, but `.include('children__children')` could be used to optimize some parts of the API.

File: osf/management/commands/populate_custom_taxonomies.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    logger.info('Successfully validated `custom`')
+    included_subjects = included_subjects | Subject.objects.filter(text__in=[map_dict['bepress'] for map_dict in customs.values()])
+    missing_subjects = Subject.objects.filter(id__in=set([hier[-1].id for ps in PreprintService.objects.filter(provider=custom_provider) for hier in ps.subject_hierarchy])).exclude(id__in=included_subjects.values_list('id', flat=True))
+    assert not missing_subjects.exists(), 'Incomplete mapping -- following subjects in use but not included:\n{}'.format(missing_subjects.all())
+    logger.info('Successfully validated mapping completeness')
+
+def create_subjects_recursive(custom_provider, root_text, exclude_texts, parent=None):
+    logger.info('Duplicating BePress subject {} on {}'.format(root_text, custom_provider._id))
+    bepress_subj = Subject.objects.get(provider=BEPRESS_PROVIDER, text=root_text)
+    custom_subj = Subject(text=root_text, parent=parent, bepress_subject=bepress_subj, provider=custom_provider)
```
Comment (cwisecarver) [inline osf/management/commands/populate_custom_taxonomies.py]: `Subject.objects.create()` would obviate the save on the next line.

File: osf/management/commands/populate_custom_taxonomies.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+def map_custom_subject(custom_provider, name, parent, mapping):
+    logger.info('Attempting to create subject {} on {} from {} with {}'.format(name, custom_provider._id, mapping, 'parent {}'.format(parent) if parent else 'no parent'))
+    if parent:
+        parent_subject = Subject.objects.filter(provider=custom_provider, text=parent).first()
+    else:
+        parent_subject = None
+    bepress_subject = Subject.objects.get(provider=BEPRESS_PROVIDER, text=mapping)
+    if parent and not parent_subject:
+        return False
+    custom_subject = Subject(provider=custom_provider, text=name, parent=parent_subject, bepress_subject=bepress_subject)
```
Comment (cwisecarver) [inline osf/management/commands/populate_custom_taxonomies.py]: `Subject.objects.create()` would obviate the save.

File: osf/models/subject.py [hunk 5]
```diff
@@ -1,25 +1,33 @@
...
-    modm_query = None
 
-    text = models.CharField(null=False, max_length=256, unique=True)  # max length on prod: 73
+    text = models.CharField(null=False, max_length=256)  # max length on prod: 73
     parent = models.ForeignKey('self', related_name='children', null=True, blank=True, on_delete=models.SET_NULL, validators=[validate_subject_hierarchy_length])
     bepress_subject = models.ForeignKey('self', related_name='aliases', null=True, blank=True, on_delete=models.deletion.CASCADE)
     provider = models.ForeignKey(PreprintProvider, related_name='subjects', on_delete=models.deletion.CASCADE)
 
+    objects = SubjectQuerySet.as_manager()
+
```
Comment (cwisecarver) [inline osf/models/subject.py]: Do you need [`_base_manager`](https://docs.djangoproject.com/en/1.11/topics/db/managers/#using-managers-for-related-object-access)?

Comment (mfraezz) [inline osf/models/subject.py]: I don't think any current behavior necessitates it, but it might avoid a headache down the line.

Comment (mfraezz) [inline osf/models/subject.py]: Resolved in https://github.com/CenterForOpenScience/osf.io/pull/7163/commits/60c380ee177c69f6da2c166e5a090d760664e9e1

File: osf/management/commands/populate_custom_taxonomies.py (hunk 3 above)
Comment (mfraezz) [inline osf/management/commands/populate_custom_taxonomies.py]: Still only one SQL statement ¯\\\_(ツ)_/¯

File: osf/management/commands/populate_custom_taxonomies.py (hunk 4 above)
Comment (mfraezz) [inline osf/management/commands/populate_custom_taxonomies.py]: > Still only one SQL statement ¯\\\_(ツ)_/¯

**Comment by mfraezz** [comment]
The management command may still need to be updated, depending on how the final input comes out, but merging to simplify migration bumping.

**Pull request merged**